In [4]:
%matplotlib inline
import pandas as pd
import numpy as np

from IPython.display import display

In [5]:
pd.options.display.min_rows = 20 # para que no lesee

# Datos

In [6]:
#datasets de autos en EEUU, para cada nuevo modelo nuevo tienen una tonelada de datos
dataset = 'https://github.com/mattharrison/datasets/raw/master/data/vehicles.csv.zip'

#leyendo
autos = pd.read_csv(dataset)


/var/folders/w6/2grt96qn39vgm51r4hz21bz40000gn/T/ipykernel_2169/1438520860.py:5: DtypeWarning: Columns (68,70,71,72,73,74,76,79) have mixed types. Specify dtype option on import or set low_memory=False.
  autos = pd.read_csv(dataset)


In [7]:
autos

barrels08  barrelsA08  charge120  charge240  city08  city08U  cityA08  \
0      15.695714         0.0        0.0        0.0      19      0.0        0   
1      29.964545         0.0        0.0        0.0       9      0.0        0   
2      12.207778         0.0        0.0        0.0      23      0.0        0   
3      29.964545         0.0        0.0        0.0      10      0.0        0   
4      17.347895         0.0        0.0        0.0      17      0.0        0   
5      14.982273         0.0        0.0        0.0      21      0.0        0   
6      13.184400         0.0        0.0        0.0      22      0.0        0   
7      13.733750         0.0        0.0        0.0      23      0.0        0   
8      12.677308         0.0        0.0        0.0      23      0.0        0   
9      13.184400         0.0        0.0        0.0      23      0.0        0   
...          ...         ...        ...        ...     ...      ...      ...   
41134  16.480500         0.0        0.0        0.0      18      0.0        0   
41135  12.677308         0.0        0.0        0.0      23      0.0        0   
41136  13.733750         0.0        0.0        0.0      21      0.0        0   
41137  11.771786         0.0        0.0        0.0      24      0.0        0   
41138  13.184400         0.0        0.0        0.0      21      0.0        0   
41139  14.982273         0.0        0.0        0.0      19      0.0        0   
41140  14.330870         0.0        0.0        0.0      20      0.0        0   
41141  15.695714         0.0        0.0        0.0      18      0.0        0   
41142  15.695714         0.0        0.0        0.0      18      0.0        0   
41143  18.311667         0.0        0.0        0.0      16      0.0        0   

       cityA08U  cityCD  cityE  ...  mfrCode  c240Dscr  charge240b  c240bDscr  \
0           0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
1           0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
2           0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
3           0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
4           0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
5           0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
6           0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
7           0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
8           0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
9           0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
...         ...     ...    ...  ...      ...       ...         ...        ...   
41134       0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
41135       0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
41136       0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
41137       0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
41138       0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
41139       0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
41140       0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
41141       0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
41142       0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
41143       0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   

                          createdOn                    modifiedOn  startStop  \
0      Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013        NaN   
1      Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013        NaN   
2      Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013        NaN   
3      Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013        NaN   
4      Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 20

In [8]:
# 41.144 filas y 83 columnas

In [9]:
autos.columns

Index(['barrels08', 'barrelsA08', 'charge120', 'charge240', 'city08',
       'city08U', 'cityA08', 'cityA08U', 'cityCD', 'cityE', 'cityUF', 'co2',
       'co2A', 'co2TailpipeAGpm', 'co2TailpipeGpm', 'comb08', 'comb08U',
       'combA08', 'combA08U', 'combE', 'combinedCD', 'combinedUF', 'cylinders',
       'displ', 'drive', 'engId', 'eng_dscr', 'feScore', 'fuelCost08',
       'fuelCostA08', 'fuelType', 'fuelType1', 'ghgScore', 'ghgScoreA',
       'highway08', 'highway08U', 'highwayA08', 'highwayA08U', 'highwayCD',
       'highwayE', 'highwayUF', 'hlv', 'hpv', 'id', 'lv2', 'lv4', 'make',
       'model', 'mpgData', 'phevBlended', 'pv2', 'pv4', 'range', 'rangeCity',
       'rangeCityA', 'rangeHwy', 'rangeHwyA', 'trany', 'UCity', 'UCityA',
       'UHighway', 'UHighwayA', 'VClass', 'year', 'youSaveSpend', 'guzzler',
       'trans_dscr', 'tCharger', 'sCharger', 'atvType', 'fuelType2', 'rangeA',
       'evMotor', 'mfrCode', 'c240Dscr', 'charge240b', 'c240bDscr',
       'createdOn', 'modifiedOn

# Tipos (*Types*)

In [10]:
cols = ['city08','comb08','highway08', 'cylinders','displ','drive','eng_dscr','fuelCost08','make','model','trany','range','createdOn','year']

In [11]:
autos[cols].dtypes

city08          int64
comb08          int64
highway08       int64
cylinders     float64
displ         float64
drive          object
eng_dscr       object
fuelCost08      int64
make           object
model          object
trany          object
range           int64
createdOn      object
year            int64
dtype: object

object no es sólo cuando es string, cuando hay muchos int y algunos nan también pandas pone object y eso ocupa ene memoria
int8, int16, int32, etc son tipos de pandas/numpy pero no de python

In [12]:
#ahí están los límites
print (np.iinfo('int64'))
print (np.iinfo('int32'))
print (np.iinfo('int16'))
print (np.iinfo('int8'))

Machine parameters for int64
---------------------------------------------------------------
min = -9223372036854775808
max = 9223372036854775807
---------------------------------------------------------------

Machine parameters for int32
---------------------------------------------------------------
min = -2147483648
max = 2147483647
---------------------------------------------------------------

Machine parameters for int16
---------------------------------------------------------------
min = -32768
max = 32767
---------------------------------------------------------------

Machine parameters for int8
---------------------------------------------------------------
min = -128
max = 127
---------------------------------------------------------------



In [13]:
#object ocupan ene memoria
autos[cols].memory_usage(deep=True)

Index             128
city08         329152
comb08         329152
highway08      329152
cylinders      329152
displ          329152
drive         3028369
eng_dscr      2135693
fuelCost08     329152
make          2606267
model         2813134
trany         2933276
range          329152
createdOn     3497240
year           329152
dtype: int64

In [14]:
#object ocupan ene memoria
print (f"Memoria : {autos[cols].memory_usage(deep=True).sum()/1e6:.2f} Mb")

Memoria : 19.65 Mb


# Enteros (*int*)

In [15]:
autos[cols].select_dtypes(int).describe() #selecciona todos los que son int

city08        comb08     highway08    fuelCost08         range  \
count  41144.000000  41144.000000  41144.000000  41144.000000  41144.000000   
mean      18.369045     20.616396     24.504667   2362.335942      0.793506   
std        7.905886      7.674535      7.730364    654.981925     13.041592   
min        6.000000      7.000000      9.000000    500.000000      0.000000   
25%       15.000000     17.000000     20.000000   1900.000000      0.000000   
50%       17.000000     20.000000     24.000000   2350.000000      0.000000   
75%       20.000000     23.000000     28.000000   2700.000000      0.000000   
max      150.000000    136.000000    124.000000   7400.000000    370.000000   

               year  
count  41144.000000  
mean    2001.535266  
std       11.142414  
min     1984.000000  
25%     1991.000000  
50%     2002.000000  
75%     2011.000000  
max     2020.000000

In [16]:
#se puede escribir así también y es más claro
(autos
 [cols]
 .select_dtypes(int)
 .describe()    
)

city08        comb08     highway08    fuelCost08         range  \
count  41144.000000  41144.000000  41144.000000  41144.000000  41144.000000   
mean      18.369045     20.616396     24.504667   2362.335942      0.793506   
std        7.905886      7.674535      7.730364    654.981925     13.041592   
min        6.000000      7.000000      9.000000    500.000000      0.000000   
25%       15.000000     17.000000     20.000000   1900.000000      0.000000   
50%       17.000000     20.000000     24.000000   2350.000000      0.000000   
75%       20.000000     23.000000     28.000000   2700.000000      0.000000   
max      150.000000    136.000000    124.000000   7400.000000    370.000000   

               year  
count  41144.000000  
mean    2001.535266  
std       11.142414  
min     1984.000000  
25%     1991.000000  
50%     2002.000000  
75%     2011.000000  
max     2020.000000

El int64 parece que es medio exagerado para todos los casos

In [17]:
#city08 que es el city milleage va desde 6 a 150 ->int16
#comb08 combined milleafe va desde 7 a 136 ->int16
#highway va desde 9 a 124 -> int8 (es un pequeño riesgo porque está cerca)
#fuelcost08 va de 500 a 7400 -> int16
#range (de autos eléctricos) va de 0 a 370 ->int16
#year va de 1984 a 2020 ->int16

In [18]:
(
autos
[cols]
.astype({'highway08':'int8', 'city08':'int16','comb08':'int16'})
.select_dtypes([int,'int8']) 
.describe()   
)

highway08    fuelCost08         range          year
count  41144.000000  41144.000000  41144.000000  41144.000000
mean      24.504667   2362.335942      0.793506   2001.535266
std        7.730364    654.981925     13.041592     11.142414
min        9.000000    500.000000      0.000000   1984.000000
25%       20.000000   1900.000000      0.000000   1991.000000
50%       24.000000   2350.000000      0.000000   2002.000000
75%       28.000000   2700.000000      0.000000   2011.000000
max      124.000000   7400.000000    370.000000   2020.000000

In [19]:
#se ve bien parece mirando los números

In [20]:
(
autos
[cols]
.astype({'highway08':'int8', 'city08':'int16','comb08':'int16','fuelCost08':'int16','range':'int16','year':'int16'})
.select_dtypes(['integer']) #esto viene de numpy son todos los que parecen int
.describe()   
)

city08        comb08     highway08    fuelCost08         range  \
count  41144.000000  41144.000000  41144.000000  41144.000000  41144.000000   
mean      18.369045     20.616396     24.504667   2362.335942      0.793506   
std        7.905886      7.674535      7.730364    654.981925     13.041592   
min        6.000000      7.000000      9.000000    500.000000      0.000000   
25%       15.000000     17.000000     20.000000   1900.000000      0.000000   
50%       17.000000     20.000000     24.000000   2350.000000      0.000000   
75%       20.000000     23.000000     28.000000   2700.000000      0.000000   
max      150.000000    136.000000    124.000000   7400.000000    370.000000   

               year  
count  41144.000000  
mean    2001.535266  
std       11.142414  
min     1984.000000  
25%     1991.000000  
50%     2002.000000  
75%     2011.000000  
max     2020.000000

In [21]:
memoria = (
autos
[cols]
.astype({'highway08':'int8', 'city08':'int16','comb08':'int16','fuelCost08':'int16','range':'int16','year':'int16'})
.memory_usage(deep=True)
.sum()
)

print (f"Memoria: {memoria/1e6:.2f} Mb")

Memoria: 18.12 Mb


In [22]:
#bajamos 1Mb un poco más como 10%

# Floats

In [23]:
(
    autos
    [cols]
    .select_dtypes(['float'])
)

cylinders  displ
0            4.0    2.0
1           12.0    4.9
2            4.0    2.2
3            8.0    5.2
4            4.0    2.2
5            4.0    1.8
6            4.0    1.8
7            4.0    1.6
8            4.0    1.6
9            4.0    1.8
...          ...    ...
41134        4.0    2.1
41135        4.0    1.9
41136        4.0    1.9
41137        4.0    1.9
41138        4.0    1.9
41139        4.0    2.2
41140        4.0    2.2
41141        4.0    2.2
41142        4.0    2.2
41143        4.0    2.2

[41144 rows x 2 columns]

In [24]:
#cilindros se ve raro, debiera ser entero
#una forma es ver si hay nan
#count te tira los que no son nan
#por lo que si lo divido por la dimensión 0 y no es 100%, entonces hay nan

autos.cylinders.count()/autos.shape[0]

0.9949931946334825

In [25]:
autos.cylinders.value_counts(dropna=False)

4.0     15938
6.0     14284
8.0      8801
5.0       771
12.0      626
3.0       279
NaN       206
10.0      170
2.0        59
16.0       10
Name: cylinders, dtype: int64

Efectivamente hay 206 Nan, me imagino autos eléctricos o algo así

In [26]:
#probemos con query que la lleva
(
    autos
    [cols]
    .query('cylinders.isna()')
)

city08  comb08  highway08  cylinders  displ              drive  \
7138       81      85         91        NaN    NaN                NaN   
7139       81      72         64        NaN    NaN      2-Wheel Drive   
8143       81      72         64        NaN    NaN      2-Wheel Drive   
8144       74      65         58        NaN    NaN                NaN   
8146       45      39         33        NaN    NaN      2-Wheel Drive   
8147       84      75         66        NaN    NaN                NaN   
9212       87      78         69        NaN    NaN      2-Wheel Drive   
9213       45      39         33        NaN    NaN      2-Wheel Drive   
10329      87      78         69        NaN    NaN      2-Wheel Drive   
21413      22      24         28        NaN    NaN      4-Wheel Drive   
...       ...     ...        ...        ...    ...                ...   
34407      73      72         71        NaN    NaN  Front-Wheel Drive   
34408     118     108         97        NaN    NaN  Front-Wheel Drive   
34409     114     104         94        NaN    NaN  Front-Wheel Drive   
34538      74      74         73        NaN    NaN    All-Wheel Drive   
34561      80      76         72        NaN    NaN      4-Wheel Drive   
34563     138     131        124        NaN    NaN   Rear-Wheel Drive   
34564     140     133        124        NaN    NaN   Rear-Wheel Drive   
34565     115     111        107        NaN    NaN    All-Wheel Drive   
34566     104     104        104        NaN    NaN    All-Wheel Drive   
34567      98      97         96        NaN    NaN    All-Wheel Drive   

      eng_dscr  fuelCost08    make                               model  \
7138       NaN         800  Nissan                            Altra EV   
7139       NaN         900  Toyota                             RAV4 EV   
8143       NaN         900  Toyota                             RAV4 EV   
8144       NaN        1000    Ford                               Th!nk   
8146       NaN        1700    Ford              Explorer USPS Electric   
8147       NaN         900  Nissan                          Hyper-Mini   
9212       NaN         850  Toyota                             RAV4 EV   
9213       NaN        1700    Ford              Explorer USPS Electric   
10329      NaN         850  Toyota                             RAV4 EV   
21413      NaN        1750  Subaru                            RX Turbo   
...        ...         ...     ...                                 ...   
34407      NaN         900     BYD                                  e6   
34408      NaN         600  Nissan        Leaf (62 kW-hr battery pack)   
34409      NaN         650  Nissan  Leaf SV/SL (62 kW-hr battery pack)   
34538      NaN         900    Audi                              e-tron   
34561      NaN         850  Jaguar                              I-Pace   
34563      NaN         500   Tesla              Model 3 Standard Range   
34564      NaN         500   Tesla         Model 3 Standard Range Plus   
34565      NaN         600   Tesla                  Model S Long Range   
34566      NaN         650   Tesla    Model S Performance (19" Wheels)   
34567      NaN         700   Tesla    Model S Performance (21" Wheels)   

                trany  range                     createdOn  year  
7138              NaN     90  Tue Jan 01 00:00:00 EST 2013  2000  
7139              NaN     88  Tue Jan 01 00:00:00 EST 2013  2000  
8143              NaN     88  Tue Jan 01 00:00:00 EST 2013  2001  
8144              NaN     29  Tue Jan 01 00:00:00 EST 2013  2001  
8146              NaN     38  Tue Jan 01 00:00:00 EST 2013  2001  
8147              NaN     33  Tue Jan 01 00:00:00 EST 2013  2001  
9212              NaN     95  Tue Jan 01 00:00:00 EST 2013  2002  
9213              NaN     38  Tue Jan 01 00:00:00 EST 2013  2002  
10329             NaN     95  Tue Jan 01 00:00:00 EST 2013  2003  
21413    Manual 5-spd      0  Tue Jan 01 00:00:00 EST 2013  1985  
...               ...    ...      

parece mayormente que son los eléctricos

In [27]:
#agreguemos cylinders y displ poniendo un cero donde es nan
#además displ como float16
(
    autos
    [cols]
    .assign(cylinders=autos.cylinders.fillna(0).astype('int8'), disp=autos.displ.fillna(0))
    .astype({'highway08':'int8', 'city08':'int16','comb08':'int16','fuelCost08':'int16','range':'int16','year':'int16'})
    .describe()
)

city08        comb08     highway08     cylinders         displ  \
count  41144.000000  41144.000000  41144.000000  41144.000000  40940.000000   
mean      18.369045     20.616396     24.504667      5.688460      3.294238   
std        7.905886      7.674535      7.730364      1.797009      1.357151   
min        6.000000      7.000000      9.000000      0.000000      0.000000   
25%       15.000000     17.000000     20.000000      4.000000      2.200000   
50%       17.000000     20.000000     24.000000      6.000000      3.000000   
75%       20.000000     23.000000     28.000000      6.000000      4.300000   
max      150.000000    136.000000    124.000000     16.000000      8.400000   

         fuelCost08         range          year          disp  
count  41144.000000  41144.000000  41144.000000  41144.000000  
mean    2362.335942      0.793506   2001.535266      3.277904  
std      654.981925     13.041592     11.142414      1.373415  
min      500.000000      0.000000   1984.000000      0.000000  
25%     1900.000000      0.000000   1991.000000      2.200000  
50%     2350.000000      0.000000   2002.000000      3.000000  
75%     2700.000000      0.000000   2011.000000      4.300000  
max     7400.000000    370.000000   2020.000000      8.400000

In [28]:
#ya no perdemos data, bkn

#los floats en su gran dimension
print (np.finfo(np.float16))
print (np.finfo(np.float32))
print (np.finfo(np.float64))

Machine parameters for float16
---------------------------------------------------------------
precision =   3   resolution = 1.00040e-03
machep =    -10   eps =        9.76562e-04
negep =     -11   epsneg =     4.88281e-04
minexp =    -14   tiny =       6.10352e-05
maxexp =     16   max =        6.55040e+04
nexp =        5   min =        -max
smallest_normal = 6.10352e-05   smallest_subnormal = 5.96046e-08
---------------------------------------------------------------

Machine parameters for float32
---------------------------------------------------------------
precision =   6   resolution = 1.0000000e-06
machep =    -23   eps =        1.1920929e-07
negep =     -24   epsneg =     5.9604645e-08
minexp =   -126   tiny =       1.1754944e-38
maxexp =    128   max =        3.4028235e+38
nexp =        8   min =        -max
smallest_normal = 1.1754944e-38   smallest_subnormal = 1.4012985e-45
---------------------------------------------------------------

Machine parameters for float64
---

In [29]:
memoria = (
    autos
    [cols]
    .assign(cylinders=autos.cylinders.fillna(0).astype('int8'), disp=autos.displ.fillna(0).astype('float16'))
    .astype({'highway08':'int8', 'city08':'int16','comb08':'int16','fuelCost08':'int16','range':'int16','year':'int16'})
    .memory_usage(deep=True)
    .sum()
)
print (f"Memoria: {memoria/1e6:.2f} Mb")

Memoria: 17.92 Mb


Ahorramos un poquito más, del orden de 2 Mb

In [30]:
#veamos la columna trany, se ve rara (transmisión)
autos.trany.value_counts(dropna=False)

Automatic 4-spd                     11047
Manual 5-spd                         8361
Automatic 3-spd                      3151
Automatic (S6)                       3106
Manual 6-spd                         2757
Automatic 5-spd                      2203
Automatic (S8)                       1665
Automatic 6-spd                      1619
Manual 4-spd                         1483
Automatic (S5)                        833
Automatic (variable gear ratios)      826
Automatic 7-spd                       724
Automatic 8-spd                       433
Automatic (AM-S7)                     424
Automatic (S7)                        327
Automatic 9-spd                       293
Automatic (AM7)                       245
Automatic (S4)                        233
Automatic (AV-S6)                     208
Automatic (A1)                        201
Automatic (AM6)                       151
Automatic (AV-S7)                     139
Automatic (S10)                       124
Automatic (AM-S6)                 

In [31]:
#esto tiene dos datos, uno que es Automatico/Manual y otro las velocidades, entonces agreguemos una nueva columna que sean los automaticos.
# por otro lado el drive se ve que es como categórico
autos.drive.value_counts()

Front-Wheel Drive             14236
Rear-Wheel Drive              13831
4-Wheel or All-Wheel Drive     6648
All-Wheel Drive                3015
4-Wheel Drive                  1460
2-Wheel Drive                   507
Part-time 4-Wheel Drive         258
Name: drive, dtype: int64

In [32]:
#en cuales falta la transmision (drive)?
(
    autos
    [cols]
    .query('drive.isna()')
)

city08  comb08  highway08  cylinders  displ drive            eng_dscr  \
7138       81      85         91        NaN    NaN   NaN                 NaN   
8144       74      65         58        NaN    NaN   NaN                 NaN   
8147       84      75         66        NaN    NaN   NaN                 NaN   
18217      18      21         25        4.0    2.0   NaN               (FFS)   
18218      20      22         26        4.0    1.5   NaN               (FFS)   
18219      13      15         20        8.0    5.7   NaN      (350 V8) (FFS)   
18220      13      15         20        8.0    5.7   NaN      (350 V8) (FFS)   
18221      15      17         20        6.0    3.0   NaN          (FFS,TRBO)   
18222      16      18         20        6.0    3.0   NaN               (FFS)   
18223      16      18         22        6.0    3.0   NaN          (FFS,TRBO)   
...       ...     ...        ...        ...    ...   ...                 ...   
20063      13      15         19        8.0    5.0   NaN      (FFS) CA model   
20064      13      15         20        8.0    5.0   NaN  (GM-OLDS) CA model   
20065      14      16         19        8.0    5.0   NaN  (GM-CHEV) CA model   
20387      14      14         15        4.0    2.4   NaN      (FFS) CA model   
21129      14      16         21        8.0    3.5   NaN  GUZZLER  FFS,TURBO   
23029      79      85         94        NaN    NaN   NaN           Lead Acid   
23030      35      37         39        NaN    NaN   NaN                NiMH   
23032      49      48         46        NaN    NaN   NaN                 NaN   
23037      49      48         46        NaN    NaN   NaN                 NaN   
23040     102      98         94        NaN    NaN   NaN                 NaN   

       fuelCost08        make                 model            trany  range  \
7138          800      Nissan              Altra EV              NaN     90   
8144         1000        Ford                 Th!nk              NaN     29   
8147          900      Nissan            Hyper-Mini              NaN     33   
18217        2000  Alfa Romeo    Spider Veloce 2000     Manual 5-spd      0   
18218        1900     Bertone                  X1/9     Manual 5-spd      0   
18219        2800   Chevrolet              Corvette  Automatic 4-spd      0   
18220        2800   Chevrolet              Corvette     Manual 4-spd      0   
18221        2500      Nissan                 300ZX  Automatic 4-spd      0   
18222        2350      Nissan                 300ZX  Automatic 4-spd      0   
18223        2350      Nissan                 300ZX     Manual 5-spd      0   
...           ...         ...                   ...              ...    ...   
20063        2800     Mercury   Grand Marquis Wagon  Automatic 4-spd      0   
20064        2800  Oldsmobile  Custom Cruiser Wagon  Automatic 4-spd      0   
20065        2650     Pontiac      Parisienne Wagon  Automatic 4-spd      0   
20387        3000      Nissan    Pickup Cab Chassis     Manual 5-spd      0   
21129        3250       Lotus             Esprit V8     Manual 5-spd      0   
23029         800         GMC                   EV1   Automatic (A1)     55   
23030        1750         GMC                   EV1   Automatic (A1)    105   
23032        1400       Honda               EV Plus   Automatic (A1)     81   
23037        1400       Honda               EV Plus   Automatic (A1)     81   
23040         650        MINI                 MiniE   Automatic (A1)    100   

                          createdOn  year  
7138   Tue Jan 01 00:00:00 EST 2013  2000  
8144   Tue Jan 01 00:00:00 EST 2013  2001  
8147   Tue Jan 01 00:00:00 EST 2013  2001  
18217  Tue Jan 01 00:00:00 EST 2013  1984  
18218  Tue Jan 01 00:00:00 EST 2013  1984  
18219  Tue Jan 01 00:00:00 EST 2013  1984  
18220  Tue Jan 01 00:00:00 EST 2013  1984  
18221  Tue Jan 01 00:00:00 EST 2013  1984  
18222  Tue Jan 01 00:00:00 EST 2013  1984  
18223  Tue Jan 01 00:00:00 EST 2013  1984  
...                    

In [33]:
#sera que los viejos no tienen?
(autos
 [cols]
 .groupby('year')
 .drive
 .nunique())

year
1984    3
1985    4
1986    4
1987    3
1988    3
1989    3
1990    3
1991    3
1992    3
1993    3
1994    3
1995    4
1996    3
1997    4
1998    4
1999    4
2000    4
2001    4
2002    4
2003    4
2004    4
2005    4
2006    4
2007    4
2008    3
2009    4
2010    6
2011    5
2012    5
2013    5
2014    5
2015    5
2016    5
2017    5
2018    5
2019    5
2020    5
Name: drive, dtype: int64

In [34]:
#parece que no, asumamos que los que no tienen drive (transmisión) son "Other" y chao

memoria = (
    autos
    [cols]
    .assign(cylinders=autos.cylinders.fillna(0).astype('int8'), 
            disp=autos.displ.fillna(0).astype('float16'),
            drive=autos.drive.fillna('Other').astype('category')) #categorizo porque son pocos
    .astype({'highway08':'int8', 'city08':'int16','comb08':'int16',
             'fuelCost08':'int16','range':'int16','year':'int16',
             'make':'category'}) #también la marca hagamosla como categoría, se que es un número limitado
    .memory_usage(deep=True)
    .sum()
)
print (f"Memoria: {memoria/1e6:.2f} Mb")

Memoria: 12.42 Mb


In [35]:
#seguimos bajando la memoria
#sobre la transmisión separemos los autos manuales porque trany dice si es Manual/Auto y el número de velocidades (lo vimos arriba)


memoria = (
    autos
    [cols]
    .assign(cylinders=autos.cylinders.fillna(0).astype('int8'), 
            disp=autos.displ.fillna(0).astype('float16'),
            drive=autos.drive.fillna('Other').astype('category'),
            automatic=autos.trany.str.contains('Auto'),                #boolean si es automático (simplismo) 
            speeds= autos.trany.str.extract(r'(\d)+').fillna(20).astype('int8'))    # regex que saca el número del texto, si no tiene asumo 20 (cualquier cosq)
    .astype({'highway08':'int8', 'city08':'int16','comb08':'int16',
             'fuelCost08':'int16','range':'int16','year':'int16',
             'make':'category'})
    .drop(columns=['trany'])  #ya no lo necesito
    .memory_usage(deep=True)
    .sum()
)
print (f"Memoria: {memoria/1e6:.2f} Mb")

Memoria: 10.96 Mb


# Fechas

In [36]:
autos[cols].createdOn.value_counts()

Tue Jan 01 00:00:00 EST 2013    34217
Fri Jun 06 00:00:00 EDT 2014      100
Thu Jul 07 00:00:00 EDT 2016      100
Thu Jul 31 00:00:00 EDT 2014       95
Wed Jun 13 00:00:00 EDT 2018       93
Tue Sep 11 00:00:00 EDT 2018       87
Tue Aug 02 00:00:00 EDT 2016       86
Fri Aug 03 00:00:00 EDT 2018       86
Thu Aug 03 00:00:00 EDT 2017       86
Wed Apr 04 00:00:00 EDT 2018       84
                                ...  
Wed Apr 10 00:00:00 EDT 2019        1
Mon Oct 13 00:00:00 EDT 2014        1
Thu Mar 28 00:00:00 EDT 2019        1
Tue Nov 15 00:00:00 EST 2016        1
Mon Jan 12 00:00:00 EST 2015        1
Mon Dec 03 00:00:00 EST 2018        1
Wed Apr 25 00:00:00 EDT 2018        1
Tue Mar 04 00:00:00 EST 2014        1
Fri Feb 28 00:00:00 EST 2014        1
Thu Jul 17 00:00:00 EDT 2014        1
Name: createdOn, Length: 269, dtype: int64

In [37]:
#la fecha está en un formato medio raro, ese EDT no lo va a entender pandas (o no se como hacerlo)
# reemplazamos por la diferencia con GMT y pasamos a datetime

memoria = (
    autos
    [cols]
    .assign(cylinders=autos.cylinders.fillna(0).astype('int8'), 
            disp=autos.displ.fillna(0).astype('float16'),
            drive=autos.drive.fillna('Other').astype('category'),
            automatic=autos.trany.str.contains('Auto'),                #boolean si es automático (simplismo) 
            speeds= autos.trany.str.extract(r'(\d)+').fillna(20).astype('int8'),
            createdOn = pd.to_datetime(autos.createdOn.replace({' EDT': '-04:00', ' EST': '-05:00'}, regex=True)))    # regex que saca el número del texto, si no tiene asumo 20 (cualquier cosq)
    .astype({'highway08':'int8', 'city08':'int16','comb08':'int16',
             'fuelCost08':'int16','range':'int16','year':'int16',
             'make':'category'})
    .drop(columns=['trany'])  #ya no lo necesito
    .memory_usage(deep=True)
    .sum()
)
print (f"Memoria: {memoria/1e6:.2f} Mb")

Memoria: 9.77 Mb


In [38]:
#veamos engine description
autos.eng_dscr.value_counts(dropna=False)

NaN                                 16153
(FFS)                                8827
SIDI                                 5526
(FFS) CA model                        926
(FFS)      (MPFI)                     734
FFV                                   701
(FFS,TRBO)                            666
(350 V8) (FFS)                        411
(GUZZLER)  (FFS)                      366
SOHC                                  354
                                    ...  
B234L/R4 (FFS,TRBO)                     1
GUZZLER V8 FFS,TURBO                    1
4.6M FFS MPFI                           1
CNG FFS                                 1
POLICE FFS MPFI                         1
B308E5 FFS,TURBO                        1
5.4E-R FFS MPFI                         1
V-6 FFS                                 1
(GUZZLER)  (FFS)      (S-CHARGE)        1
R-ENG (FFS,TRBO)                        1
Name: eng_dscr, Length: 558, dtype: int64

In [39]:
#esto si supiera habría más trabajo, faltan 16 mil. para el ejercicio tomemos las que dicen FFS como indicador y chao pescao

memoria = (
    autos
    [cols]
    .assign(cylinders=autos.cylinders.fillna(0).astype('int8'), 
            disp=autos.displ.fillna(0).astype('float16'),
            drive=autos.drive.fillna('Other').astype('category'),
            automatic=autos.trany.str.contains('Auto'),                
            speeds= autos.trany.str.extract(r'(\d)+').fillna(20).astype('int8'),
            createdOn = pd.to_datetime(autos.createdOn.replace({' EDT': '-04:00', ' EST': '-05:00'}, regex=True)),
            ffs=autos.eng_dscr.str.contains('FFS'))    
    .astype({'highway08':'int8', 'city08':'int16','comb08':'int16',
             'fuelCost08':'int16','range':'int16','year':'int16',
             'make':'category'})
    .drop(columns=['trany','eng_dscr'])  #ya no lo necesito
    .memory_usage(deep=True)
    .sum()
)
print (f"Memoria: {memoria/1e6:.2f} Mb")

Memoria: 9.01 Mb


In [40]:
#ahora está lista la pega, lo reducimos a menos de la mitad
# cuando haces chaining es mejor después hacerlo con una función que se puede arreglar después y no llenar con cambios y cosas raras
# asi si tengo que hacer un cambio lo hago acpa y no miro ochocientas lineas

def tweak_autos(autos):
    cols = ['city08','comb08','highway08', 'cylinders','displ','drive','eng_dscr','fuelCost08','make','model','trany','range','createdOn','year']
    return (
    autos
    [cols]
    .assign(cylinders=autos.cylinders.fillna(0).astype('int8'), 
            disp=autos.displ.fillna(0).astype('float16'),
            drive=autos.drive.fillna('Other').astype('category'),
            automatic=autos.trany.str.contains('Auto'),                
            speeds= autos.trany.str.extract(r'(\d)+').fillna(20).astype('int8'),
            createdOn = pd.to_datetime(autos.createdOn.replace({' EDT': '-04:00', ' EST': '-05:00'}, regex=True)),
            ffs=autos.eng_dscr.str.contains('FFS'))    
    .astype({'highway08':'int8', 'city08':'int16','comb08':'int16',
             'fuelCost08':'int16','range':'int16','year':'int16',
             'make':'category'})
    .drop(columns=['trany','eng_dscr'])  #ya no lo necesito
    )

tweak_autos(autos) #voilá

city08  comb08  highway08  cylinders  displ  \
0          19      21         25          4    2.0   
1           9      11         14         12    4.9   
2          23      27         33          4    2.2   
3          10      11         12          8    5.2   
4          17      19         23          4    2.2   
5          21      22         24          4    1.8   
6          22      25         29          4    1.8   
7          23      24         26          4    1.6   
8          23      26         31          4    1.6   
9          23      25         30          4    1.8   
...       ...     ...        ...        ...    ...   
41134      18      20         24          4    2.1   
41135      23      26         33          4    1.9   
41136      21      24         30          4    1.9   
41137      24      28         33          4    1.9   
41138      21      25         32          4    1.9   
41139      19      22         26          4    2.2   
41140      20      23         28          4    2.2   
41141      18      21         24          4    2.2   
41142      18      21         24          4    2.2   
41143      16      18         21          4    2.2   

                            drive  fuelCost08        make  \
0                Rear-Wheel Drive        2000  Alfa Romeo   
1                Rear-Wheel Drive        3850     Ferrari   
2               Front-Wheel Drive        1550       Dodge   
3                Rear-Wheel Drive        3850       Dodge   
4      4-Wheel or All-Wheel Drive        2700      Subaru   
5               Front-Wheel Drive        1900      Subaru   
6               Front-Wheel Drive        1700      Subaru   
7               Front-Wheel Drive        1750      Toyota   
8               Front-Wheel Drive        1600      Toyota   
9               Front-Wheel Drive        1700      Toyota   
...                           ...         ...         ...   
41134           Front-Wheel Drive        2100        Saab   
41135           Front-Wheel Drive        1600      Saturn   
41136           Front-Wheel Drive        1750      Saturn   
41137           Front-Wheel Drive        1500      Saturn   
41138           Front-Wheel Drive        1700      Saturn   
41139           Front-Wheel Drive        1900      Subaru   
41140           Front-Wheel Drive        1850      Subaru   
41141  4-Wheel or All-Wheel Drive        2000      Subaru   
41142  4-Wheel or All-Wheel Drive        2000      Subaru   
41143  4-Wheel or All-Wheel Drive        2900      Subaru   

                     model  range                  createdOn  year      disp  \
0       Spider Veloce 2000      0  2013-01-01 00:00:00-05:00  1985  2.000000   
1               Testarossa      0  2013-01-01 00:00:00-05:00  1985  4.898438   
2                  Charger      0  2013-01-01 00:00:00-05:00  1985  2.199219   
3      B150/B250 Wagon 2WD      0  2013-01-01 00:00:00-05:00  1985  5.199219   
4         Legacy AWD Turbo      0  2013-01-01 00:00:00-05:00  1993  2.199219   
5                   Loyale      0  2013-01-01 00:00:00-05:00  1993  1.799805   
6                   Loyale      0  2013-01-01 00:00:00-05:00  1993  1.799805   
7                  Corolla      0  2013-01-01 00:00:00-05:00  1993  1.599609   
8                  Corolla      0  2013-01-01 00:00:00-05:00  1993  1.599609   
9                  Corolla      0  2013-01-01 00:00:00-05:00  1993  1.799805   
...                    ...    ...                        ...   ...       ...   
41134                  900      0  2013-01-01 00:00:00-05:00  1993  2.099609   
41135                   SL      0  2013-01-01 00:00:00-05:00  1993  1.900391   
41136                   SL      0  2013-01-01 00:00:00-05:00  1993  1.900391   
41137                   SL      0  2013-01-01 00:00:00-05:00  1993  1.900391   
41138                   SL      0  2013-01-01 00:00:00-05:00  1993  1.900391   
41139               Legacy      0  2013-01-01 00:00:00-05:00  1993  2.199219   
41140               Legacy      0  

# Chain o Flow Programming
Ev vez de usar variables intermedias uno ocupa el hecho de que la mayoría de las operaciones retornan un objeto y así se lee como una colección de recetas con pasos

Si uno no encuentra la forma de hacer chain se puede usar .pipe

In [41]:
def tweak_autos(autos):
    cols = ['city08','comb08','highway08', 'cylinders','displ','drive','eng_dscr','fuelCost08','make','model','trany','range','createdOn','year']
    return (
    autos
    [cols]
    .assign(cylinders=autos.cylinders.fillna(0).astype('int8'), 
            disp=autos.displ.fillna(0).astype('float16'),
            drive=autos.drive.fillna('Other').astype('category'),
            automatic=autos.trany.str.contains('Auto'),                
            speeds= autos.trany.str.extract(r'(\d)+').fillna(20).astype('int8'),
            createdOn = pd.to_datetime(autos.createdOn.replace({' EDT': '-04:00', ' EST': '-05:00'}, regex=True)),
            ffs=autos.eng_dscr.str.contains('FFS'))    
    .astype({'highway08':'int8', 'city08':'int16','comb08':'int16',
             'fuelCost08':'int16','range':'int16','year':'int16',
             'make':'category'})
    .drop(columns=['trany','eng_dscr'])  #ya no lo necesito
    )


Comparemos con este desastre

In [42]:
a1 = autos[cols]
cyls=autos.cylinders.fillna(0)
cyls2=cyls.astype('int8')
a1['cylinders'] = cyls2
#ETC ETC ETC

/var/folders/w6/2grt96qn39vgm51r4hz21bz40000gn/T/ipykernel_2169/1905911837.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  a1['cylinders'] = cyls2


Si necesitamos mostrar algo se puede hacer pipe

In [43]:
from IPython.display import display

def get_var(df, var_name):
    globals()[var_name] = df
    return df

def tweak_autos(autos):
    cols = ['city08','comb08','highway08', 'cylinders','displ','drive','eng_dscr','fuelCost08','make','model','trany','range','createdOn','year']
    return (
    autos
    [cols]
    .pipe(get_var,"df3") #crea variable df3 es un truco y puedo verlo como variable de paso intermedio
    .assign(cylinders=autos.cylinders.fillna(0).astype('int8'), 
            disp=autos.displ.fillna(0).astype('float16'),
            drive=autos.drive.fillna('Other').astype('category'),
            automatic=autos.trany.str.contains('Auto'),                
            speeds= autos.trany.str.extract(r'(\d)+').fillna(20).astype('int8'),
            createdOn = pd.to_datetime(autos.createdOn.replace({' EDT': '-04:00', ' EST': '-05:00'}, regex=True)),
            ffs=autos.eng_dscr.str.contains('FFS'))    
    .astype({'highway08':'int8', 'city08':'int16','comb08':'int16',
             'fuelCost08':'int16','range':'int16','year':'int16',
             'make':'category'})
    .drop(columns=['trany','eng_dscr'])  
    )

tweak_autos(autos)
df3 #acá está mostrando df3

city08  comb08  highway08  cylinders  displ  \
0          19      21         25        4.0    2.0   
1           9      11         14       12.0    4.9   
2          23      27         33        4.0    2.2   
3          10      11         12        8.0    5.2   
4          17      19         23        4.0    2.2   
5          21      22         24        4.0    1.8   
6          22      25         29        4.0    1.8   
7          23      24         26        4.0    1.6   
8          23      26         31        4.0    1.6   
9          23      25         30        4.0    1.8   
...       ...     ...        ...        ...    ...   
41134      18      20         24        4.0    2.1   
41135      23      26         33        4.0    1.9   
41136      21      24         30        4.0    1.9   
41137      24      28         33        4.0    1.9   
41138      21      25         32        4.0    1.9   
41139      19      22         26        4.0    2.2   
41140      20      23         28        4.0    2.2   
41141      18      21         24        4.0    2.2   
41142      18      21         24        4.0    2.2   
41143      16      18         21        4.0    2.2   

                            drive          eng_dscr  fuelCost08        make  \
0                Rear-Wheel Drive             (FFS)        2000  Alfa Romeo   
1                Rear-Wheel Drive         (GUZZLER)        3850     Ferrari   
2               Front-Wheel Drive             (FFS)        1550       Dodge   
3                Rear-Wheel Drive               NaN        3850       Dodge   
4      4-Wheel or All-Wheel Drive        (FFS,TRBO)        2700      Subaru   
5               Front-Wheel Drive             (FFS)        1900      Subaru   
6               Front-Wheel Drive             (FFS)        1700      Subaru   
7               Front-Wheel Drive             (FFS)        1750      Toyota   
8               Front-Wheel Drive             (FFS)        1600      Toyota   
9               Front-Wheel Drive             (FFS)        1700      Toyota   
...                           ...               ...         ...         ...   
41134           Front-Wheel Drive             (FFS)        2100        Saab   
41135           Front-Wheel Drive  (TBI)      (FFS)        1600      Saturn   
41136           Front-Wheel Drive  (MFI)      (FFS)        1750      Saturn   
41137           Front-Wheel Drive  (TBI)      (FFS)        1500      Saturn   
41138           Front-Wheel Drive  (MFI)      (FFS)        1700      Saturn   
41139           Front-Wheel Drive             (FFS)        1900      Subaru   
41140           Front-Wheel Drive             (FFS)        1850      Subaru   
41141  4-Wheel or All-Wheel Drive             (FFS)        2000      Subaru   
41142  4-Wheel or All-Wheel Drive             (FFS)        2000      Subaru   
41143  4-Wheel or All-Wheel Drive        (FFS,TRBO)        2900      Subaru   

                     model            trany  range  \
0       Spider Veloce 2000     Manual 5-spd      0   
1               Testarossa     Manual 5-spd      0   
2                  Charger     Manual 5-spd      0   
3      B150/B250 Wagon 2WD  Automatic 3-spd      0   
4         Legacy AWD Turbo     Manual 5-spd      0   
5                   Loyale  Automatic 3-spd      0   
6                   Loyale     Manual 5-spd      0   
7                  Corolla  Automatic 3-spd      0   
8                  Corolla     Manual 5-spd      0   
9                  Corolla  Automatic 4-spd      0   
...                    ...              ...    ...   
41134                  900     Manual 5-spd      0   
41135                   SL  Automatic 4-spd      0   
41136                   SL  Automatic 4-spd      0   
41137                   SL     Manual 5-spd      0   
41138                   SL     Manual 5-spd      0   
41139               Legacy  Automatic 4-spd      0   
41140               Legacy     Manual 5-spd      0   
41141           Legacy AWD  Automatic 4-spd      0   

In [44]:
from IPython.display import display

def get_var(df, var_name):
    globals()[var_name] = df
    return df

def tweak_autos(autos):
    cols = ['city08','comb08','highway08', 'cylinders','displ','drive','eng_dscr','fuelCost08','make','model','trany','range','createdOn','year']
    return (
    autos
    [cols]
    .pipe(get_var,"df3") #crea variable df3 es un truco
    .assign(cylinders=autos.cylinders.fillna(0).astype('int8'), 
            disp=autos.displ.fillna(0).astype('float16'),
            drive=autos.drive.fillna('Other').astype('category'),
            automatic=autos.trany.str.contains('Auto'),                
            speeds= autos.trany.str.extract(r'(\d)+').fillna(20).astype('int8'),
            createdOn = pd.to_datetime(autos.createdOn.replace({' EDT': '-04:00', ' EST': '-05:00'}, regex=True)),
            ffs=autos.eng_dscr.str.contains('FFS'))
    .pipe(lambda df:display(df) or df) #esto es para debug y display es para notebook y lo muestra 
    #jupyter por defecto solo muestra el ultimo paso esto permite ver el intermedio   
    .astype({'highway08':'int8', 'city08':'int16','comb08':'int16',
             'fuelCost08':'int16','range':'int16','year':'int16',
             'make':'category'})
    .drop(columns=['trany','eng_dscr'])  
    )

tweak_autos(autos)


city08  comb08  highway08  cylinders  displ  \
0          19      21         25          4    2.0   
1           9      11         14         12    4.9   
2          23      27         33          4    2.2   
3          10      11         12          8    5.2   
4          17      19         23          4    2.2   
5          21      22         24          4    1.8   
6          22      25         29          4    1.8   
7          23      24         26          4    1.6   
8          23      26         31          4    1.6   
9          23      25         30          4    1.8   
...       ...     ...        ...        ...    ...   
41134      18      20         24          4    2.1   
41135      23      26         33          4    1.9   
41136      21      24         30          4    1.9   
41137      24      28         33          4    1.9   
41138      21      25         32          4    1.9   
41139      19      22         26          4    2.2   
41140      20      23         28          4    2.2   
41141      18      21         24          4    2.2   
41142      18      21         24          4    2.2   
41143      16      18         21          4    2.2   

                            drive          eng_dscr  fuelCost08        make  \
0                Rear-Wheel Drive             (FFS)        2000  Alfa Romeo   
1                Rear-Wheel Drive         (GUZZLER)        3850     Ferrari   
2               Front-Wheel Drive             (FFS)        1550       Dodge   
3                Rear-Wheel Drive               NaN        3850       Dodge   
4      4-Wheel or All-Wheel Drive        (FFS,TRBO)        2700      Subaru   
5               Front-Wheel Drive             (FFS)        1900      Subaru   
6               Front-Wheel Drive             (FFS)        1700      Subaru   
7               Front-Wheel Drive             (FFS)        1750      Toyota   
8               Front-Wheel Drive             (FFS)        1600      Toyota   
9               Front-Wheel Drive             (FFS)        1700      Toyota   
...                           ...               ...         ...         ...   
41134           Front-Wheel Drive             (FFS)        2100        Saab   
41135           Front-Wheel Drive  (TBI)      (FFS)        1600      Saturn   
41136           Front-Wheel Drive  (MFI)      (FFS)        1750      Saturn   
41137           Front-Wheel Drive  (TBI)      (FFS)        1500      Saturn   
41138           Front-Wheel Drive  (MFI)      (FFS)        1700      Saturn   
41139           Front-Wheel Drive             (FFS)        1900      Subaru   
41140           Front-Wheel Drive             (FFS)        1850      Subaru   
41141  4-Wheel or All-Wheel Drive             (FFS)        2000      Subaru   
41142  4-Wheel or All-Wheel Drive             (FFS)        2000      Subaru   
41143  4-Wheel or All-Wheel Drive        (FFS,TRBO)        2900      Subaru   

                     model            trany  range                  createdOn  \
0       Spider Veloce 2000     Manual 5-spd      0  2013-01-01 00:00:00-05:00   
1               Testarossa     Manual 5-spd      0  2013-01-01 00:00:00-05:00   
2                  Charger     Manual 5-spd      0  2013-01-01 00:00:00-05:00   
3      B150/B250 Wagon 2WD  Automatic 3-spd      0  2013-01-01 00:00:00-05:00   
4         Legacy AWD Turbo     Manual 5-spd      0  2013-01-01 00:00:00-05:00   
5                   Loyale  Automatic 3-spd      0  2013-01-01 00:00:00-05:00   
6                   Loyale     Manual 5-spd      0  2013-01-01 00:00:00-05:00   
7                  Corolla  Automatic 3-spd      0  2013-01-01 00:00:00-05:00   
8                  Corolla     Manual 5-spd      0  2013-01-01 00:00:00-05:00   
9                  Corolla  Automatic 4-spd      0  2013-01-01 00:00:00-05:00   
...                    ...              ...    ...                        ...   
41134                  900     Manual 5-spd      0  2013-01-01 00:00:00-05:00   
41135                   SL

city08  comb08  highway08  cylinders  displ  \
0          19      21         25          4    2.0   
1           9      11         14         12    4.9   
2          23      27         33          4    2.2   
3          10      11         12          8    5.2   
4          17      19         23          4    2.2   
5          21      22         24          4    1.8   
6          22      25         29          4    1.8   
7          23      24         26          4    1.6   
8          23      26         31          4    1.6   
9          23      25         30          4    1.8   
...       ...     ...        ...        ...    ...   
41134      18      20         24          4    2.1   
41135      23      26         33          4    1.9   
41136      21      24         30          4    1.9   
41137      24      28         33          4    1.9   
41138      21      25         32          4    1.9   
41139      19      22         26          4    2.2   
41140      20      23         28          4    2.2   
41141      18      21         24          4    2.2   
41142      18      21         24          4    2.2   
41143      16      18         21          4    2.2   

                            drive  fuelCost08        make  \
0                Rear-Wheel Drive        2000  Alfa Romeo   
1                Rear-Wheel Drive        3850     Ferrari   
2               Front-Wheel Drive        1550       Dodge   
3                Rear-Wheel Drive        3850       Dodge   
4      4-Wheel or All-Wheel Drive        2700      Subaru   
5               Front-Wheel Drive        1900      Subaru   
6               Front-Wheel Drive        1700      Subaru   
7               Front-Wheel Drive        1750      Toyota   
8               Front-Wheel Drive        1600      Toyota   
9               Front-Wheel Drive        1700      Toyota   
...                           ...         ...         ...   
41134           Front-Wheel Drive        2100        Saab   
41135           Front-Wheel Drive        1600      Saturn   
41136           Front-Wheel Drive        1750      Saturn   
41137           Front-Wheel Drive        1500      Saturn   
41138           Front-Wheel Drive        1700      Saturn   
41139           Front-Wheel Drive        1900      Subaru   
41140           Front-Wheel Drive        1850      Subaru   
41141  4-Wheel or All-Wheel Drive        2000      Subaru   
41142  4-Wheel or All-Wheel Drive        2000      Subaru   
41143  4-Wheel or All-Wheel Drive        2900      Subaru   

                     model  range                  createdOn  year      disp  \
0       Spider Veloce 2000      0  2013-01-01 00:00:00-05:00  1985  2.000000   
1               Testarossa      0  2013-01-01 00:00:00-05:00  1985  4.898438   
2                  Charger      0  2013-01-01 00:00:00-05:00  1985  2.199219   
3      B150/B250 Wagon 2WD      0  2013-01-01 00:00:00-05:00  1985  5.199219   
4         Legacy AWD Turbo      0  2013-01-01 00:00:00-05:00  1993  2.199219   
5                   Loyale      0  2013-01-01 00:00:00-05:00  1993  1.799805   
6                   Loyale      0  2013-01-01 00:00:00-05:00  1993  1.799805   
7                  Corolla      0  2013-01-01 00:00:00-05:00  1993  1.599609   
8                  Corolla      0  2013-01-01 00:00:00-05:00  1993  1.599609   
9                  Corolla      0  2013-01-01 00:00:00-05:00  1993  1.799805   
...                    ...    ...                        ...   ...       ...   
41134                  900      0  2013-01-01 00:00:00-05:00  1993  2.099609   
41135                   SL      0  2013-01-01 00:00:00-05:00  1993  1.900391   
41136                   SL      0  2013-01-01 00:00:00-05:00  1993  1.900391   
41137                   SL      0  2013-01-01 00:00:00-05:00  1993  1.900391   
41138                   SL      0  2013-01-01 00:00:00-05:00  1993  1.900391   
41139               Legacy      0  2013-01-01 00:00:00-05:00  1993  2.199219   
41140               Legacy      0  

También puedo comentar una linea y ver como se mueve, de a poco y no pasa nada

# Don't mutate
Esto es cuando uno uso inplace=True
* No hay beneficios de rendimiento
* No se puede hacer chaining
* SettingWithCopyWarning fun

# Don't Apply
Apply es básicamente un loop, es mega lento

In [45]:
def tweak_autos(autos):
    cols = ['city08','comb08','highway08', 'cylinders','displ','drive','eng_dscr','fuelCost08','make','model','trany','range','createdOn','year']
    return (
    autos
    [cols]
    .assign(cylinders=autos.cylinders.fillna(0).astype('int8'), 
            disp=autos.displ.fillna(0).astype('float16'),
            drive=autos.drive.fillna('Other').astype('category'),
            automatic=autos.trany.str.contains('Auto'),                
            speeds= autos.trany.str.extract(r'(\d)+').fillna(20).astype('int8'),
            createdOn = pd.to_datetime(autos.createdOn.replace({' EDT': '-04:00', ' EST': '-05:00'}, regex=True)),
            ffs=autos.eng_dscr.str.contains('FFS'))
    .astype({'highway08':'int8', 'city08':'int16','comb08':'int16',
             'fuelCost08':'int16','range':'int16','year':'int16',
             'make':'category'})
    .drop(columns=['trany','eng_dscr'])  
    )


In [49]:
#tratemos de ver el ahorro en litros/100 kms
def to_lper100km(val):
    return 235.215/val

autos2 = tweak_autos(autos)


In [50]:
%timeit autos2.city08.apply(to_lper100km)

4.43 ms ± 14.7 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [52]:
#esto da el mismo resultado
%timeit 235.215/autos2.city08

44.9 µs ± 179 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [53]:
#es mucho más rápido, casi 100 veces en este caso

In [54]:
def is_american(val):
    return val  in {'Chevrolet', 'Ford','Dodge', 'GMC', 'Tesla'}

In [55]:
%%timeit
autos2.make.apply(is_american)

166 µs ± 1.85 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [56]:
%%timeit
autos2.make.isin({'Chevrolet', 'Ford','Dodge', 'GMC', 'Tesla'})

372 µs ± 3.74 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


In [57]:
%%timeit
#strings en pandas no son muy optimizadas entonces las string operations son lentos
autos2.make.astype(str).isin({'Chevrolet', 'Ford','Dodge', 'GMC', 'Tesla'})

5.17 ms ± 49.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [58]:
def country(val):
    if val in {'Chevrolet', 'Ford','Dodge', 'GMC', 'Tesla'}:
        return 'US'
    return 'Other'

In [59]:
%%timeit
(autos2
 .assign(country=autos2.make.apply(country))
 )

3.13 ms ± 71.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [60]:
%%timeit
values = {'Chevrolet', 'Ford','Dodge', 'GMC', 'Tesla'}
(autos2
 .assign(country='US')
 .assign(country=lambda df_:df_.country.where(df_.make.isin(values),'Other'))
 )

7.35 ms ± 58 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [61]:
#where sirve pero es medio extraño cuando hay muchas variables o ifs
#acá es más lento por los strings


In [63]:
%%timeit
values = {'Chevrolet', 'Ford','Dodge', 'GMC', 'Tesla'}

#una buena opcion es usar np.select
(autos2
 .assign(country=np.select([autos2.make.isin(values)],['US'],'Other'))
 )


4.89 ms ± 50.7 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


# Master Aggregation